wasm2c/template/wasm-rt-mem-impl.tmpl.c

```diff
@@ -0,0 +1,286 @@
+  return succeeded ? 0 : -1;
+}
+
+static int os_mprotect(void* addr, size_t size) {
+  if (size == 0) {
+    return 0;
+  }
+  void* ret = VirtualAlloc(addr, size, MEM_COMMIT, PAGE_READWRITE);
+  if (ret == addr) {
+    return 0;
+  }
+  VirtualFree(addr, 0, MEM_RELEASE);
+  return -1;
+}
+
+static void os_print_last_error(const char* msg) {
+  DWORD errorMessageID = GetLastError();
+  if (errorMessageID != 0) {
+    LPSTR messageBuffer = 0;
+    // The api creates the buffer that holds the message
+    size_t size = FormatMessageA(
+        FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_SYSTEM |
+            FORMAT_MESSAGE_IGNORE_INSERTS,
+        NULL, errorMessageID, MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT),
+        (LPSTR)&messageBuffer, 0, NULL);
+    (void)size;
+    printf("%s. %s\n", msg, messageBuffer);
+    LocalFree(messageBuffer);
+  } else {
+    printf("%s. No error code.\n", msg);
+  }
+}
+
+#else
+static void* os_mmap(size_t size) {
+  int map_prot = PROT_NONE;
+  int map_flags = MAP_ANONYMOUS | MAP_PRIVATE;
+  uint8_t* addr = mmap(NULL, size, map_prot, map_flags, -1, 0);
+  if (addr == MAP_FAILED)
+    return NULL;
+  return addr;
+}
+
+static int os_munmap(void* addr, size_t size) {
+  return munmap(addr, size);
+}
+
+static int os_mprotect(void* addr, size_t size) {
+  return mprotect(addr, size, PROT_READ | PROT_WRITE);
+}
+
+static void os_print_last_error(const char* msg) {
+  perror(msg);
+}
+
+#endif
+
+#if WASM_RT_USE_MMAP
+
+static uint64_t get_alloc_size_for_mmap(
+    wasm_rt_TEMPLATE_REPLACE_memory_t* memory) {
+  assert(!memory->is64 &&
+         "memory64 is not yet compatible with WASM_RT_USE_MMAP");
+#if WASM_RT_MEMCHECK_GUARD_PAGES
+  /* Reserve 8GiB. */
+  const uint64_t max_size = 0x200000000ul;
+  return max_size;
+#else
+  if (memory->max_pages != 0) {
+    const uint64_t max_size = memory->max_pages * PAGE_SIZE;
+    return max_size;
+  }
+
+  /* Reserve 4GiB. */
+  const uint64_t max_size = 0x100000000ul;
+  return max_size;
+#endif
+}
+
+#endif
+
+void wasm_rt_allocate_memory_TEMPLATE_REPLACE(
+    wasm_rt_TEMPLATE_REPLACE_memory_t* memory,
+    uint64_t initial_pages,
+    uint64_t max_pages,
+    bool is64) {
+  uint64_t byte_length = initial_pages * PAGE_SIZE;
+  memory->size = byte_length;
+  memory->pages = initial_pages;
+  memory->max_pages = max_pages;
+  memory->is64 = is64;
+  MEMORY_LOCK_VAR_INIT(memory->mem_lock);
+
+#if WASM_RT_USE_MMAP
+  const uint64_t mmap_size = get_alloc_size_for_mmap(memory);
+  void* addr = os_mmap(mmap_size);
+  if (!addr) {
+    os_print_last_error("os_mmap failed.");
+    abort();
+  }
+  int ret = os_mprotect(addr, byte_length);
+  if (ret != 0) {
+    os_print_last_error("os_mprotect failed.");
+    abort();
+  }
+  memory->data = addr;
+#else
+  memory->data = calloc(byte_length, 1);
+#endif
+}
+
+static uint64_t grow_memory_impl(wasm_rt_TEMPLATE_REPLACE_memory_t* memory,
+                                 uint64_t delta) {
```

Comment:
It looks like only a few function in this file actually use `_TEMPLATE_REPLACE_`  .. I wonder if its worth leaving everything else in the non-template file?

Also, I wonder if we could simply use the pre-processor here rather than cmake processing?

How about something like this where we just re-include the same file for each type?

```
#define MEMORY_TYPE wasm_rt_memory_t
#include <wasm_rt_memfuncs.inc>
#define MEMORY_TYPE wasm_rt_shared_memory_t
#include <wasm_rt_memfuncs.inc>
```

This could avoid the cmake stuff and the new checked-in prebuilt?

Reply:
Yup, I had this earlier, but @keithw  suggested that the template approach is preferred. Personally, I have no preference and I am happy to go with the concensus

Reply:
@keithw WDYT?  I think it we can use the simple c pre-processor approach for this kind of thing thats better than cmake magic.

Reply:
My only real preference would be to try to keep the approach consistent between "what we do for tables" (funcref + externref) vs. "what we do for memories" (normal + shared). I don't care a ton about what the approach is and don't want to apply stop energy, but it would be nice to avoid a flowering of different mechanisms. If you want to use a multiply included `.inc` file for normal+shared memories, fine with me, but then let's do the same thing for funcref+externref tables.

Reply:
Are you saying that we have already take on of these approaches for (funcref + externref) ?     Can you point me to that?

Reply:
Was thinking of https://github.com/WebAssembly/wabt/blob/main/src/template/wasm2c.declarations.c#L554-L604

Reply:
Will fix this in a follow up PR